### PR TITLE
Add pformat to config object

### PR DESCRIPTION
**Pull request by danr**
I sometimes find `pprint.pformat` suboptimal so it would be nice if I could drop-in replace it for snoop. This PR adds `pp_pformat` to the config object (and the `snoop.install` method) to override it with some other function.

It is also useful if you want to configure `pprint.pformat`, for example passing `sort_keys=False`.

### CODE REVIEW

**Comment by alexmojaki** [comment]
Thanks, I'm in favour of this. Something similar was requested in https://github.com/alexmojaki/snoop/issues/13.

Please:

1. Change the name from `pp_pformat` to just `pformat` which I think is easier.
2. Set the default in `Config.__init__` rather than `write`.
3. Use `pprintpp` if it can be imported, i.e. https://github.com/wolever/pprintpp is installed.
4. Otherwise, try importing [`prettyprinter`](https://github.com/tommikaikkonen/prettyprinter). Let me know if you think the precedence should be swapper, they look like similar quality.
5. Add some documentation.

**Comment by alexmojaki** [comment]
Oh and there's also https://github.com/panyanyany/beeprint. Why are they all equally good looking lol

**Comment by danr** [comment]
Thanks for the feedback. The `prettyprinter` library seems like the highest quality from a quick glance.

The precedence I expect is: 
1. The user has explicitly passed `pformat` to the install method or config object
2. If installed: `prettyprinter.pformat`.
3. If installed: `pprintpp.pformat`.
4. The standard `pprint.pformat`.

I'm a bit hesitant to this check for external libraries during runtime but it's your shot of course. Personally I will just be using my own pretty printer for now.

On the failing test case: I'm not sure what happens on 3.4, but all of the versions get really confused after the `with snoop:` block. Should I design the test a bit differently perhaps?

**Comment by danr** [comment]
New commit pushed now. I am not sure how to test if the 3rd party printing libraries are there with the tox setup. Ideally both with and without 3rd party printing libraries should be tested.

**Comment by danr** [comment]
I wrote:
> On the failing test case: I'm not sure what happens on 3.4, but all of the versions get really confused after the `with snoop:` block. Should I design the test a bit differently perhaps?
I designed the test a bit differently, now this distractor is not present.

**Comment by danr** [comment]
Ok, python 3.4 is still confused. Is this to be expected?

**Comment by alexmojaki** [comment]
`pp` is not expected to work in 3.4.

Use `FIX_SNOOP_TESTS=1 tox` to generate the sample files.

File: tests/samples/pp_custom_pformat.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+
+def test():
+    x = 10
+    y = 20
+    assert pp(pp(x + y) + y) == 50
+    assert pp.deep(lambda: (x + y) + y) == 50
+    pp([14, 15, 16])
+
+    with snoop:
+        z = 30  # custom pformat is not used here
```
Comment (alexmojaki) [inline tests/samples/pp_custom_pformat.py:10]: Does this add anything to the test?

**Comment by alexmojaki** [comment]
> all of the versions get really confused after the with snoop: block.

That really shouldn't happen, I will try finding out why later on.

File: tests/samples/pp_custom_pformat.py (hunk 1 above)
Comment (danr) [reply tests/samples/pp_custom_pformat.py:10]: Yes and no :) If we start using `pformat` for `snoop` this test will remind us of this change.

**Comment by danr** [comment]
Do you have some way to install all the python versions for tox? I'm installing `python34` from arch linux aur now...

**Comment by alexmojaki** [comment]
`pyenv` is great, for me at least, not sure about ubuntu.

If you have difficulties, I'll handle the tests.

**Comment by danr** [comment]
Hmm yes please do, I'm struggling getting a 3.4 working.

**Comment by alexmojaki** [comment]
> The precedence I expect is:
> 
> 1. The user has explicitly passed `pformat` to the install method or config object
> 2. If installed: `prettyprinter.pformat`.
> 3. If installed: `pprintpp.pformat`.
> 4. The standard `pprint.pformat`.
> 
> I'm a bit hesitant to this check for external libraries during runtime but it's your shot of course. Personally I will just be using my own pretty printer for now.

Add https://github.com/panyanyany/beeprint to that list.

Install all the libraries for tests. Test that it uses `prettyprinter` by default. Set `sys.modules["prettyprinter"] = None` (I haven't actually tested this but I'm sure something similar must be possible). Test that it now uses `pprintpp` by default, etc.

**Comment by alexmojaki** [comment]
Oh, beeprint is not just like pprint, forget that.

**Comment by alexmojaki** [comment]
I'm catching all exceptions rather than just ImportError because that's been an issue for me before: https://github.com/alexmojaki/birdseye/pull/82

**Comment by alexmojaki** [comment]
> > all of the versions get really confused after the with snoop: block.
> 
> That really shouldn't happen, I will try finding out why later on.

Fixed in https://github.com/alexmojaki/snoop/pull/34. This was a very weird and interesting problem, see https://github.com/cool-RR/PySnooper/issues/214 for the explanation. Thanks for your contribution which picked this up.

**Comment by danr** [comment]
Added a few lines of documentation and tests for the 3rd party libraries. Turns out you can simply do `sys.modules['pprintpp'] = {}` to make `from pprintpp import pformat` fail (since that empty dictionary does not contain `{}`). 

> Fixed in #34. This was a very weird and interesting problem, see cool-RR/PySnooper#214 for the explanation. Thanks for your contribution which picked this up.

Nice!

**Comment by danr** [comment]
I only ran tests for 2.7 and 3.9 since that's the versions I have around :grimacing:

File: snoop/configuration.py [hunk 2]
```diff
@@ -62,8 +64,9 @@ def install(
...
         - `function_qualname`: The qualified name of the current function.
-        
-        If you want a custom column, please open an issue to tell me what you're interested in! In the meantime, you can pass a list, where the elements are either strings or callables. The callables should take one argument, which will be an `Event` object. It has attributes `frame`, `event`, and `arg`, as specified in [`sys.settrace()`](https://docs.python.org/3/library/sys.html#sys.settrace), and other attributes which may change. 
+
+        If you want a custom column, please open an issue to tell me what you're interested in! In the meantime, you can pass a list, where the elements are either strings or callables. The callables should take one argument, which will be an `Event` object. It has attributes `frame`, `event`, and `arg`, as specified in [`sys.settrace()`](https://docs.python.org/3/library/sys.html#sys.settrace), and other attributes which may change.
+    - `pformat`: set the pretty formatting function `pp` uses. Default is to use the first of `prettyprinter.pformat`, `pprintpp.pformat` and `pprint.pformat` that can be imported.
```
Comment (alexmojaki) [inline snoop/configuration.py:69]: Please also mention the pformat argument here and under Output configuration.

**Comment by alexmojaki** [comment]
Had to take a detour to fix birdseye which was broken by a new release of sqlalchemy...

File: snoop/configuration.py (hunk 2 above)
Comment (danr) [reply snoop/configuration.py:69]: It is mentioned there (?) I pushed a new commit which adds it to the README's Output configuration section.

Comment (alexmojaki) [reply snoop/configuration.py:69]: lol my brain only saw mention of the libraries and not the argument.

**Comment by alexmojaki** [comment]
Ah, I think I meant to say that pformat needs to be mentioned here: https://github.com/alexmojaki/snoop#pp---awesome-print-debugging

Anyway, this is released under 0.3.0

**Comment by danr** [comment]
Thanks Alex!

**Pull request merged**